include/openthread/tcp.h

```diff
@@ -0,0 +1,521 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread TCP API.
+ *
+ */
+
+#ifndef OPENTHREAD_TCP_H_
+#define OPENTHREAD_TCP_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-tcp
+ *
+ * @brief
+ *   This module includes functions that control TCP communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This a vectored packet structure for use with TCP.
+ *
+ * A single @p otLinkedBuffer structure references an array of bytes in memory,
+ * via @p mData and @p mLength. The @p mNext field is used to form a chain of
+ * @p otLinkedBuffer structures.
+ */
+typedef struct otLinkedBuffer
+{
+    uint8_t* mData;              ///< Pointer to data referenced by this linked buffer
+    size_t mLength;              ///< Length of this linked buffer
+    struct otLinkedBuffer *mNext; ///< Pointer to the next linked buffer in this chain, or NULL if this is the end of the chain
+} otLinkedBuffer;
+
+/**
+ * This callback informs the application that the TCP 3-way handshake is
+ * complete and that the connection is now established.
+ */
+typedef void (*otTcpConnectDone)(void *aContext);
```

Comment:
I followed the same pattern as the UDP Socket API. For UDP sockets, what OpenThread currently does is provide the context pointer associated with the socket in the callback, rather than a pointer to the socket itself. That's what the `aContext` argument is doing here.

I think that we should stick to using `aContext` rather than providing a pointer to the socket directly, at the very least because that's what OpenThread's UDP socket seems to do. I think the reason the UDP API does this is because a different context (e.g., some other struct which in turn contains a pointer to the socket) might be more useful to the application developer in some cases. But, I am happy to be convinced otherwise if others feel strongly that we should switch to a pointer to the socket for TCP. :)

Reply:
I think adding socket pointer would be helpful and make it easier for API users. 

> That said, I'm open to discussing this further if anyone feels strongly that we should be passing pointers to sockets instead of pointers to the associated context.

We should still keep `aContext` as well (all callbacks should include a `aContext` argument which should pass back what the callback provider gave when it registered the callback). 

Regarding `aContext` use in OT, it is generally tied to a callback and its meaning/use is for callback provider  (i.e. API user) to decide. 

The common use pattern is that the `aContext` is set to a pointer to the class object that is making the call into OT.  If the calling object owns/manages multiple TCP sockets, it then cannot use this simple model and need to either assign different callbacks for each socket or define a a different context struct to track the socket.

Regarding UDP socket, in my view, it would be better for it to also include the socket but we keep it as if due to legacy/compatibility reason. 

----

In particular related to the use of context here in this callback
```
typedef void (*otTcpAcceptDone)(void *aPassiveContext, void *aActiveContext, const otSockAddr *aPeer)
```
I think we want the callback definition to have a single `aContext`. Using `aPassiveContext` and `aActiveContext` this way is indirectly assuming that context are associated with some socket struct (which I don't think OT API should assume this in general case, from OT API definitions contexts are associated with a callback). I think make sense to pass the socket pointer directly here.

Reply:
Thanks @abtink. I think those are good reasons. Can you clarify what the signature of the `otTcpAcceptDone` would be? I'm thinking that, if we incorporate your suggestions, it would look something like this:
```
typedef void (*otTcpAcceptDone)(void *aPassiveContext, otTcpPassiveSocket *aPassiveSocket, void *aActiveContext, otTcpActiveSocket *aActiveSocket, const otSockAddr *aPeer);
```
Is this what you had in mind? I think keeping both contexts around is a good idea since (in principle) the passive socket and active socket could have different contexts, and the application developer might benefit from both.

I'm fine going through all of the callbacks and adding in a pointer to the socket itself as the second argument, after the context. The part I found convincing in @abtink's argument above is the observation that the context pointer is often a pointer to the class object that contains the socket, making it hard to disambiguate which socket it refers to. Before I make the change, I'd like to see if there's a broad consensus, particularly from others who are familiar with OpenThread's design patterns. I know @simonlingoogle was suggesting similar changes. @jwhui, what do you think?

Reply:
(While we wait for others to chime in, I might draft a commit with this change, which we can revert if we find later that others don't agree with it.)

Reply:
Sounds good. Thanks.

Let me share some info on some design patterns that we use in OT:

In OT API design pattern, a (`void *`) context is generally associated with a callback pointer (not necessarily any other related objects). 
- When a callback is registered/set, the caller provides the callback fn pointer along with `aContext` value.
- OT ensures to pass back the same `aContext` as an argument (typically first argument) when in invokes the the callback function pointer.

The main intention/use-case of this model is to make it look as if we calling a method on an object using `C` API , i.e. `aContext` kind of acts as `this` pointer in a C++ method call.

So `otTcpAcceptDone()` as a callback can have a single `aContext` associated with it. Having `aPassiveContext` and `aActiveContext` in the callback breaks this model and assumes that the contexts are associated with socket objects and not the callback. I think we should continue the same model unless we have a good justification/reason for changing it. 

So here is  format I'd suggest for the `otTcpAcceptDone()`:
```
typedef void (*otTcpAcceptDone)(void *aContext, otTcpPassiveSocket *aPassiveSocket, otTcpActiveSocket *aActiveSocket, const otSockAddr *aPeer);
```

Reply:
OK, thanks for clarifying that. I'll update the callback to the signature that you suggested.

To build on the idea of a context associated with a callback rather than a socket --- does this mean that each callback should have its own context?

Reply:
We can have different callbacks each with their own context, but I think if we register a bunch of callbacks all together then it may make sense to use the same `aContext` for them all. 

I think we have examples of both cases in existing OT APIs, for example in CoAP module.

In `coap.h` we have:
https://github.com/openthread/openthread/blob/e3b971e497bb047d44bc1c9722a81c521354a7f4/include/openthread/coap.h#L907-L915
where we have three callbacks `aHandler`, `aTransmitHook`, and `aReceiveHook` all part of same API and all share same `aContext`. 

And we have "connected" callback which is registered separately and can have its own separate context.
https://github.com/openthread/openthread/blob/e3b971e497bb047d44bc1c9722a81c521354a7f4/include/openthread/coap_secure.h#L329-L340

------------

For TCP, we can consider using `otTcpActiveOpen()` as the main (and only) way to set all the callbacks which then share the same context. The question if there is any value/usecase to have separate API to be able to change the callbacks afterwards. I cannot think of any immediate benefit for it but we can consider its uses.

Reply:
I think it makes sense to have both a context pointer and socket pointer for the reasons @abtink mentioned above. In retrospect, the UDP API should have probably followed a similar model.

Reply:
Thanks @abtink and @jwhui for your feedback. I made this change to all of the callback functions in the latest update that I just pushed. I think we can resolve this conversation, but I'll let you take a look before I take that step.

Reply:
Thanks @samkumar. Looks great.